File: cpp/practise/remove-all-nodes-which-lie-on-a-path-having-sum-less-than-k.cpp
#include<bits/stdc++.h>

using namespace std;


struct Node{

    int key;
    Node *left;
    Node *right;

    Node(int data){
        key = data;
        left = NULL;
        right = NULL;
    }

};


void inOrder(Node *root){

    if(root==NULL)
        return;

    inOrder(root->left);
    cout<<root->key<<" ";
    inOrder(root->right);

}

Node* prune(Node *root, int k){

    if(root == NULL) return NULL;

    root->left = prune(root->left, k-root->key);
    root->right = prune(root->right, k-root->key);

    if(root->left == NULL && root->right== NULL){
        if(root->key< k){
            delete root;
            return NULL;
        }
    }

    return root;

}


int main(){

    int k = 45;
    struct Node *root = new Node(1);
    root->left = new Node(2);
    root->right = new Node(3);
    root->left->left = new Node(4);
    root->left->right = new Node(5);
    root->right->left = new Node(6);
    root->right->right = new Node(7);
    root->left->left->left = new Node(8);
    root->left->left->right = new Node(9);
    root->left->right->left = new Node(12);
    root->right->right->left = new Node(10);
    root->right->right->left->right = new Node(11);
    root->left->left->right->left = new Node(13);
    root->left->left->right->right = new Node(14);
    root->left->left->right->right->left = new Node(15);


    cout<<"Before Pruning :"<<endl;
    inOrder(root);


    prune(root, k);

    cout<<"After Pruning :"<<endl;
    inOrder(root);


    return 0;
}

File: py/remove-all-nodes-which-lie-on-a-path-having-sum-less-than-k.py
class Node(object):
    """docstring for Node."""
    def __init__(self, data):
        #super(Node, self).__init__()
        self.key = data
        self.left = None
        self.right = None

def inOrder(root):
    if root is None:
        return

    inOrder(root.left)
    print root.key,
    inOrder(root.right)

def prune(root, K):

    if root is None:
        return None

    root.left = prune(root.left, K-root.key)
    root.right = prune(root.right, K-root.key)

    if root.left is None and root.right is None:
        if root.key< K:
            return None

    return root

if __name__ == '__main__':
    root = Node(1)
    root.left = Node(2)
    root.right = Node(3)
    root.left.left = Node(4)
    root.left.right = Node(5)
    root.right.left = Node(6)
    root.right.right = Node(7)
    root.left.left.left = Node(8)
    root.left.left.right = Node(9)
    root.left.right.left = Node(12)
    root.right.right.left = Node(10)
    root.right.right.left.right = Node(11)
    root.left.left.right.left = Node(13)
    root.left.left.right.right = Node(14)
    root.left.left.right.right.left = Node(15)


    print 'Before Pruning :'
    inOrder(root)

    K = 45
    prune(root, K)

    print '\n After Pruning : '
    inOrder(root)
